Enable bluetooth for MIMXRT1040-EVK

lylezhu2012 opened the pull request:
Add Bluetooth support for MIMXRT1040-EVK. Currently, only support Murata 2EL M.2 module.

Add MIMXRT1040-EVK board support to sample handsfree/handsfree_ag.

lylezhu2012 committed:
Sample: bluetooth: add mimxrt1040_evk support

Enable mimxrt1040_evk for handsfree and
handsfree_ag samples.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

jori-nordic commented:
Thank you.
Do you have a TLDR guide for flashing this to the devkit (right out of the box, virgin)?
`west flash` fails for me with

```
jon@jori-pc:~/repos/zephyrproject/zephyr/samples/bluetooth/handsfree$ west flash 
-- west flash: rebuilding
ninja: no work to do.
-- west flash: using runner jlink
-- runners.jlink: reset after flashing requested
-- runners.jlink: JLink version: 7.80c
-- runners.jlink: Flashing file: /home/jon/repos/zephyrproject/zephyr/samples/bluetooth/handsfree/build/zephyr/zephyr.bin
FATAL ERROR: command exited with status 1: /opt/SEGGER/JLink_V780c/JLinkExe -nogui 1 -if swd -speed auto -device MIMXRT1042XXX6B -CommanderScript /tmp/tmpyrqhoifjjlink/runner.jlink -nogui 1
```

lylezhu2012 commented:
> Thank you. Do you have a TLDR guide for flashing this to the devkit (right out of the box, virgin)? `west flash` fails for me with
> 
> ```
> jon@jori-pc:~/repos/zephyrproject/zephyr/samples/bluetooth/handsfree$ west flash -i 753002278
> -- west flash: rebuilding
> ninja: no work to do.
> -- west flash: using runner jlink
> -- runners.jlink: reset after flashing requested
> -- runners.jlink: JLink version: 7.94e
> -- runners.jlink: Flashing file: /home/jon/repos/zephyrproject/zephyr/samples/bluetooth/handsfree/build/zephyr/zephyr.bin
> FATAL ERROR: command exited with status 1: /opt/SEGGER/JLink_V794e/JLinkExe -USB 753002278 -nogui 1 -if swd -speed auto -device MIMXRT1042XXX6B -CommanderScript /tmp/tmpt8hfjt0rjlink/runner.jlink -nogui 1
> ```

Usually, I use Ozone to flash the image. And I just tried `west flash` on my local side. It worked. My OS is windows 11.

lylezhu2012 committed:
boards: mimxrt1040_evk: enable Bluetooth

Add pinmux pinmux_lpuart3_flowcontrol
and pinmux_lpuart3_sleep for HCI-UART.

Add a node m2_hci_uart for HCI_UART.

And add chosen `zephyr,bt-hci =
&bt_hci_uart;`

Update section `Supported Features`,
section `Connections and IOs` and
section `Serial Port` for this feature.

Add section `Fetch Binary Blobs` to
guide how to fetch binary blobs for
BT modules.

add section `Bluetooth Module` for
hardware rework.

Signed-off-by: Lyle Zhu <lyle.zhu@nxp.com>

lylezhu2012 commented:
Fix conflict. And add reminder for debugger connection or flash failure.

jori-nordic commented:
Okay, managed to get it booting and initializing the `bluetooth/handsfree` sample. But I don't see the device when scanning with my and @theob-pro 's phones.

```
jon@jori-pc:~/repos/zephyrproject/zephyr/samples/bluetooth/handsfree$ miniterm /dev/ttyACM2 115200
--- Miniterm on /dev/ttyACM2  115200,8,N,1 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
*** Booting Zephyr OS build v3.6.0-6131-gd66fbdb3f69c ***
Bluetooth initialized
BR/EDR set connectable and discoverable done
```

For future me (and others), here's what I had to do:
- Download the EVK datasheet. Have to create NXP account for that (cmon!)
- Do the SMD rework mentioned in the "Wifi module" section: disconnect `R497, R498, R456 and R457`, connect `R96, R93`. Finding those on the PCB is left as an exercise to the reader.
- Disconnect J80 (small jumper next to the electret microphone)
- Try to flash with `west flash`. Doesn't work, it tries to use JLink but the FW shipped on the EVK is CMSIS-DAP.
- Try to flash with `west flash -r pyocd`. Still doesn't work, CPU is too new, not supported by pyOCD yet
- Update JLink + Ozone binaries to latest
- Disconnect J9 and J10. Connect an external Segger J-Link programmer (big connector on the right)
- Flash with `west flash -i [external-segger-serial]`. Now works. Flashing with Ozone also works.
- Boot. Bluetooth doesn't initialize.
- Swap out the module shipped with the EVK with the one recommended by @lylezhu2012 : `EAR00409` from embedded-artists
- Boot. Bluetooth initializes, sample goes into BR discoverable mode 🥳

lylezhu2012 commented:
> Okay, managed to get it booting and initializing the `bluetooth/handsfree` sample. But I don't see the device when scanning with my and @theob-pro 's phones.
> 
> ```
> jon@jori-pc:~/repos/zephyrproject/zephyr/samples/bluetooth/handsfree$ miniterm /dev/ttyACM2 115200
> --- Miniterm on /dev/ttyACM2  115200,8,N,1 ---
> --- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
> *** Booting Zephyr OS build v3.6.0-6131-gd66fbdb3f69c ***
> Bluetooth initialized
> BR/EDR set connectable and discoverable done
> ```
> 
> For future me (and others), here's what I had to do:
> 
> * Download the EVK datasheet. Have to create NXP account for that (cmon!)
> * Do the SMD rework mentioned in the "Wifi module" section: disconnect `R497, R498, R456 and R457`, connect `R96, R93`. Finding those on the PCB is left as an exercise to the reader.
> * Disconnect J80 (small jumper next to the electret microphone)
> * Try to flash with `west flash`. Doesn't work, it tries to use JLink but the FW shipped on the EVK is CMSIS-DAP.
> * Try to flash with `west flash -r pyocd`. Still doesn't work, CPU is too new, not supported by pyOCD yet
> * Update JLink + Ozone binaries to latest
> * Disconnect J9 and J10. Connect an external Segger J-Link programmer (big connector on the right)
> * Flash with `west flash -i [external-segger-serial]`. Now works. Flashing with Ozone also works.
> * Boot. Bluetooth doesn't initialize.
> * Swap out the module shipped with the EVK with the one recommended by @lylezhu2012 : `EAR00409` from embedded-artists
> * Boot. Bluetooth initializes, sample goes into BR discoverable mode 🥳

I think your example works. TX power of Bluetooth module is not regulated. Please bring your phone closer to the board (The distance should be less than 50cm).

jori-nordic commented:
The RSSI is not the issue.
So I copied the overlay to `peripheral_hr`, and I'm able to connect to the EVK no problem.

Maybe I broke the Classic stack with my TX rework PR? Could you test `handsfree` after rebasing on main, just to be sure?

lylezhu2012 commented:
> The RSSI is not the issue. My phone is less than 10cm away from the module. So I copied the overlay to `peripheral_hr`, and I'm able to connect to the EVK no problem.
> 
> Maybe I broke the Classic stack with my TX rework PR? Could you test `handsfree` after rebasing on main, just to be sure?

Sure. I have tested the PR with handsfree sample based on commit 012520d2d7e8d2bbc914bb5735bf1563abd0e4ea. My phone (Android OS) can connect with the evk and no issue found.

jori-nordic commented:
Anyway, thanks for the PR. It's one other platform I can test with :)
